File: app/models/preventive_maintenance.py

from sqlalchemy import String, DateTime, Enum as SAEnum, ForeignKey, Text, Numeric, Boolean, Integer
from sqlalchemy.orm import Mapped, mapped_column, relationship
from enum import Enum
from datetime import datetime
from typing import List, Optional
from app.database import Base
from app.core.datetime_utils import now_sp
from app.models.procurement import PurchaseProduct


class MaintenanceType(str, Enum):
    PREVENTIVA = "Preventiva"
    PREDITIVA = "Preditiva"
    INSPECAO = "Inspeção"
    CALIBRACAO = "Calibração"
    LUBRIFICACAO = "Lubrificação"
    LIMPEZA = "Limpeza"
    ATUALIZACAO = "Atualização"
    CORRETIVA = "Corretiva"
    PERSONALIZADA = "Personalizada"


class MaintenancePeriodicity(str, Enum):
    DIARIA = "Diária"
    SEMANAL = "Semanal"
    QUINZENAL = "Quinzenal"
    MENSAL = "Mensal"
    BIMESTRAL = "Bimestral"
    TRIMESTRAL = "Trimestral"
    SEMESTRAL = "Semestral"
    ANUAL = "Anual"
    PERSONALIZADA = "Personalizada"


class MaintenancePriority(str, Enum):
    BAIXA = "Baixa"
    MEDIA = "Média"
    ALTA = "Alta"
    URGENTE = "Urgente"


class MaintenanceCriticality(str, Enum):
    BAIXA = "Baixa"
    MEDIA = "Média"
    ALTA = "Alta"
    CRITICA = "Crítica"


class OrderStatus(str, Enum):
    ABERTA = "Aberta"
    AGENDADA = "Agendada"
    EM_ANDAMENTO = "Em andamento"
    AGUARDANDO_PECA = "Aguardando peça"
    PAUSADA = "Pausada"
    CONCLUIDA = "Concluída"
    CANCELADA = "Cancelada"


class PhotoType(str, Enum):
    ANTES = "Antes"
    DURANTE = "Durante"
    DEPOIS = "Depois"


class MaintenancePlan(Base):
    __tablename__ = "maintenance_plans"

    id: Mapped[int] = mapped_column(primary_key=True, index=True)
    nome: Mapped[str] = mapped_column(String(200), nullable=False, index=True)
    codigo: Mapped[str] = mapped_column(String(50), unique=True, index=True, nullable=False)
    descricao: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    tipo: Mapped[MaintenanceType] = mapped_column(SAEnum(MaintenanceType), default=MaintenanceType.PREVENTIVA)
    periodicidade: Mapped[MaintenancePeriodicity] = mapped_column(SAEnum(MaintenancePeriodicity), default=MaintenancePeriodicity.MENSAL)
    dias_personalizado: Mapped[Optional[int]] = mapped_column(Integer, nullable=True)
    tempo_estimado_horas: Mapped[Optional[float]] = mapped_column(Numeric(5, 2), nullable=True)
    criticidade: Mapped[MaintenanceCriticality] = mapped_column(SAEnum(MaintenanceCriticality), default=MaintenanceCriticality.MEDIA)
    prioridade: Mapped[MaintenancePriority] = mapped_column(SAEnum(MaintenancePriority), default=MaintenancePriority.MEDIA)
    ativo: Mapped[bool] = mapped_column(Boolean, default=True)
    
    # Responsável
    responsavel_id: Mapped[Optional[int]] = mapped_column(ForeignKey("users.id"), nullable=True)
    
    # Setor
    departamento_id: Mapped[Optional[int]] = mapped_column(ForeignKey("departamentos.id"), nullable=True)
    
    # Categoria (para aplicar a todos os ativos de uma categoria)
    categoria_id: Mapped[Optional[int]] = mapped_column(ForeignKey("asset_categories.id"), nullable=True)
    
    # Datas
    data_criacao: Mapped[datetime] = mapped_column(DateTime, default=now_sp)
    data_ultima_execucao: Mapped[Optional[datetime]] = mapped_column(DateTime, nullable=True)
    proxima_execucao: Mapped[datetime] = mapped_column(DateTime, nullable=False)
    
    # Relacionamentos
    responsavel = relationship("User")
    departamento = relationship("Departamento")
    categoria = relationship("AssetCategory")
    assets = relationship("MaintenancePlanAsset", back_populates="plan", cascade="all, delete-orphan")
    checklists = relationship("MaintenanceChecklist", back_populates="plan", cascade="all, delete-orphan")
    orders = relationship("MaintenanceOrder", back_populates="plan", cascade="all, delete-orphan")
    notifications = relationship("MaintenanceNotification", back_populates="plan", cascade="all, delete-orphan")


class MaintenancePlanAsset(Base):
    __tablename__ = "maintenance_plan_assets"

    id: Mapped[int] = mapped_column(primary_key=True, index=True)
    plan_id: Mapped[int] = mapped_column(ForeignKey("maintenance_plans.id"), nullable=False)
    asset_id: Mapped[int] = mapped_column(ForeignKey("assets.id"), nullable=False)
    
    # Relacionamentos
    plan = relationship("MaintenancePlan", back_populates="assets")
    asset = relationship("Asset")


class MaintenanceChecklist(Base):
    __tablename__ = "maintenance_checklists"

    id: Mapped[int] = mapped_column(primary_key=True, index=True)
    plan_id: Mapped[int] = mapped_column(ForeignKey("maintenance_plans.id"), nullable=False)
    nome: Mapped[str] = mapped_column(String(200), nullable=False)
    ordem: Mapped[int] = mapped_column(Integer, default=0)
    
    # Relacionamentos
    plan = relationship("MaintenancePlan", back_populates="checklists")
    items = relationship("MaintenanceChecklistItem", back_populates="checklist", cascade="all, delete-orphan")


class MaintenanceChecklistItem(Base):
    __tablename__ = "maintenance_checklist_items"

    id: Mapped[int] = mapped_column(primary_key=True, index=True)
    checklist_id: Mapped[int] = mapped_column(ForeignKey("maintenance_checklists.id"), nullable=False)
    descricao: Mapped[str] = mapped_column(Text, nullable=False)
    obrigatorio: Mapped[bool] = mapped_column(Boolean, default=True)
    ordem: Mapped[int] = mapped_column(Integer, default=0)
    requer_foto: Mapped[bool] = mapped_column(Boolean, default=False)
    
    # Relacionamentos
    checklist = relationship("MaintenanceChecklist", back_populates="items")


class MaintenanceOrder(Base):
    __tablename__ = "maintenance_orders"

    id: Mapped[int] = mapped_column(primary_key=True, index=True)
    numero: Mapped[str] = mapped_column(String(50), unique=True, index=True, nullable=False)
    plan_id: Mapped[Optional[int]] = mapped_column(ForeignKey("maintenance_plans.id"), nullable=True)
    asset_id: Mapped[Optional[int]] = mapped_column(ForeignKey("assets.id"), nullable=True)
    infra_predial_servico: Mapped[Optional[str]] = mapped_column(String(255), nullable=True)
    
    # Responsáveis
    tecnico_id: Mapped[Optional[int]] = mapped_column(ForeignKey("users.id"), nullable=True)
    solicitante_id: Mapped[Optional[int]] = mapped_column(ForeignKey("users.id"), nullable=True)
    
    # Dados da ordem
    status: Mapped[OrderStatus] = mapped_column(SAEnum(OrderStatus), default=OrderStatus.ABERTA)
    prioridade: Mapped[MaintenancePriority] = mapped_column(SAEnum(MaintenancePriority), default=MaintenancePriority.MEDIA)
    criticidade: Mapped[MaintenanceCriticality] = mapped_column(SAEnum(MaintenanceCriticality), default=MaintenanceCriticality.MEDIA)
    tipo: Mapped[MaintenanceType] = mapped_column(SAEnum(MaintenanceType), default=MaintenanceType.PREVENTIVA)
    
    # Datas e horários
    data_abertura: Mapped[datetime] = mapped_column(DateTime, default=now_sp)
    data_agendada: Mapped[Optional[datetime]] = mapped_column(DateTime, nullable=True)
    data_inicio: Mapped[Optional[datetime]] = mapped_column(DateTime, nullable=True)
    data_pausa: Mapped[Optional[datetime]] = mapped_column(DateTime, nullable=True)
    data_conclusao: Mapped[Optional[datetime]] = mapped_column(DateTime, nullable=True)
    tempo_total_minutos: Mapped[Optional[int]] = mapped_column(Integer, nullable=True)
    
    # Observações
    observacoes: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    solucao: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    
    # Custos
    custo_total: Mapped[Optional[float]] = mapped_column(Numeric(10, 2), nullable=True, default=0.0)
    
    # Relacionamentos com Service Desk
    service_ticket_id: Mapped[Optional[int]] = mapped_column(ForeignKey("service_tickets.id"), nullable=True)
    
    # Relacionamentos
    plan = relationship("MaintenancePlan", back_populates="orders")
    asset = relationship("Asset")
    tecnico = relationship("User", foreign_keys=[tecnico_id])
    solicitante = relationship("User", foreign_keys=[solicitante_id])
    service_ticket = relationship("ServiceTicket")
    executions = relationship("MaintenanceExecution", back_populates="order", cascade="all, delete-orphan")
    materials = relationship("MaintenanceMaterial", back_populates="order", cascade="all, delete-orphan")
    photos = relationship("MaintenancePhoto", back_populates="order", cascade="all, delete-orphan")
    history = relationship("MaintenanceHistory", back_populates="order", cascade="all, delete-orphan")
    notifications = relationship("MaintenanceNotification", back_populates="order", cascade="all, delete-orphan")

    @property
    def display_tipo(self) -> str:
        if self.tipo.value == "personalizada" and self.observacoes and self.observacoes.startswith("[TIPO: "):
            import re
            match = re.match(r'^\[TIPO: ([^\]]+)\]', self.observacoes)
            if match:
                return match.group(1)
        return self.tipo.value


class MaintenanceExecution(Base):
    __tablename__ = "maintenance_executions"

    id: Mapped[int] = mapped_column(primary_key=True, index=True)
    order_id: Mapped[int] = mapped_column(ForeignKey("maintenance_orders.id"), nullable=False)
    checklist_item_id: Mapped[int] = mapped_column(ForeignKey("maintenance_checklist_items.id"), nullable=False)
    concluido: Mapped[bool] = mapped_column(Boolean, default=False)
    observacao: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    data_execucao: Mapped[Optional[datetime]] = mapped_column(DateTime, nullable=True)
    executado_por_id: Mapped[Optional[int]] = mapped_column(ForeignKey("users.id"), nullable=True)
    
    # Relacionamentos
    order = relationship("MaintenanceOrder", back_populates="executions")
    checklist_item = relationship("MaintenanceChecklistItem")
    executado_por = relationship("User")


class MaintenanceMaterial(Base):
    __tablename__ = "maintenance_materials"

    id: Mapped[int] = mapped_column(primary_key=True, index=True)
    order_id: Mapped[int] = mapped_column(ForeignKey("maintenance_orders.id"), nullable=False)
    product_id: Mapped[Optional[int]] = mapped_column(ForeignKey("purchase_products.id"), nullable=True)
    produto: Mapped[str] = mapped_column(String(200), nullable=False)
    quantidade: Mapped[float] = mapped_column(Numeric(10, 2), nullable=False)
    valor_unitario: Mapped[float] = mapped_column(Numeric(10, 2), nullable=False)
    valor_total: Mapped[float] = mapped_column(Numeric(10, 2), nullable=False)
    observacao: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    
    # Relacionamentos
    order = relationship("MaintenanceOrder", back_populates="materials")
    product = relationship("PurchaseProduct")


class MaintenancePhoto(Base):
    __tablename__ = "maintenance_photos"

    id: Mapped[int] = mapped_column(primary_key=True, index=True)
    order_id: Mapped[int] = mapped_column(ForeignKey("maintenance_orders.id"), nullable=False)
    execution_id: Mapped[Optional[int]] = mapped_column(ForeignKey("maintenance_executions.id"), nullable=True)
    tipo: Mapped[PhotoType] = mapped_column(SAEnum(PhotoType), default=PhotoType.DURANTE)
    caminho_arquivo: Mapped[str] = mapped_column(String(255), nullable=False)
    descricao: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    data_upload: Mapped[datetime] = mapped_column(DateTime, default=now_sp)
    upload_por_id: Mapped[Optional[int]] = mapped_column(ForeignKey("users.id"), nullable=True)
    
    # Relacionamentos
    order = relationship("MaintenanceOrder", back_populates="photos")
    execution = relationship("MaintenanceExecution")
    upload_por = relationship("User")


class MaintenanceHistory(Base):
    __tablename__ = "maintenance_history"

    id: Mapped[int] = mapped_column(primary_key=True, index=True)
    order_id: Mapped[int] = mapped_column(ForeignKey("maintenance_orders.id"), nullable=False)
    acao: Mapped[str] = mapped_column(String(100), nullable=False)
    descricao: Mapped[str] = mapped_column(Text, nullable=False)
    usuario_id: Mapped[Optional[int]] = mapped_column(ForeignKey("users.id"), nullable=True)
    data_hora: Mapped[datetime] = mapped_column(DateTime, default=now_sp)
    status_anterior: Mapped[Optional[str]] = mapped_column(String(50), nullable=True)
    status_novo: Mapped[Optional[str]] = mapped_column(String(50), nullable=True)
    
    # Relacionamentos
    order = relationship("MaintenanceOrder", back_populates="history")
    usuario = relationship("User")


class MaintenanceNotification(Base):
    __tablename__ = "maintenance_notifications"

    id: Mapped[int] = mapped_column(primary_key=True, index=True)
    order_id: Mapped[Optional[int]] = mapped_column(ForeignKey("maintenance_orders.id", ondelete="CASCADE"), nullable=True)
    plan_id: Mapped[Optional[int]] = mapped_column(ForeignKey("maintenance_plans.id", ondelete="CASCADE"), nullable=True)
    usuario_id: Mapped[int] = mapped_column(ForeignKey("users.id"), nullable=False)
    tipo: Mapped[str] = mapped_column(String(50), nullable=False)
    mensagem: Mapped[str] = mapped_column(Text, nullable=False)
    lida: Mapped[bool] = mapped_column(Boolean, default=False)
    data_criacao: Mapped[datetime] = mapped_column(DateTime, default=now_sp)
    
    # Relacionamentos
    order = relationship("MaintenanceOrder", back_populates="notifications")
    plan = relationship("MaintenancePlan", back_populates="notifications")
    usuario = relationship("User")


class CustomMaintenanceType(Base):
    """Tipos de manutenção customizados (além dos tipos fixos do Enum MaintenanceType)."""
    __tablename__ = "custom_maintenance_types"

    id: Mapped[int] = mapped_column(primary_key=True, index=True)
    nome: Mapped[str] = mapped_column(String(200), nullable=False, unique=True)
    descricao: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    criado_em: Mapped[datetime] = mapped_column(DateTime, default=now_sp)

File: backend/internal/models/preventive_maintenance.go
package models

import "time"

// Maintenance type values (display values, matching Python MaintenanceType)
const (
	MaintTypePreventiva    = "Preventiva"
	MaintTypePreditiva     = "Preditiva"
	MaintTypeInspecao      = "Inspeção"
	MaintTypeCalibracao    = "Calibração"
	MaintTypeLubrificacao  = "Lubrificação"
	MaintTypeLimpeza       = "Limpeza"
	MaintTypeAtualizacao   = "Atualização"
	MaintTypeCorretiva     = "Corretiva"
	MaintTypePersonalizada = "Personalizada"
)

// Periodicity values
const (
	PeriodicidadeDiaria        = "Diária"
	PeriodicidadeSemanal       = "Semanal"
	PeriodicidadeQuinzenal     = "Quinzenal"
	PeriodicidadeMensal        = "Mensal"
	PeriodicidadeBimestral     = "Bimestral"
	PeriodicidadeTrimestral    = "Trimestral"
	PeriodicidadeSemestral     = "Semestral"
	PeriodicidadeAnual         = "Anual"
	PeriodicidadePersonalizada = "Personalizada"
)

// Priority values
const (
	PriorityBaixa   = "Baixa"
	PriorityMedia   = "Média"
	PriorityAlta    = "Alta"
	PriorityUrgente = "Urgente"
)

// Criticality values
const (
	CriticalityBaixa   = "Baixa"
	CriticalityMedia   = "Média"
	CriticalityAlta    = "Alta"
	CriticalityCritica = "Crítica"
)

// PM order status values
const (
	PMStatusAberta         = "Aberta"
	PMStatusAgendada       = "Agendada"
	PMStatusEmAndamento    = "Em andamento"
	PMStatusAguardandoPeca = "Aguardando peça"
	PMStatusPausada        = "Pausada"
	PMStatusConcluida      = "Concluída"
	PMStatusCancelada      = "Cancelada"
)

// Photo type values
const (
	PhotoAntes   = "Antes"
	PhotoDurante = "Durante"
	PhotoDepois  = "Depois"
)

type MaintenancePlan struct {
	ID                 uint     `gorm:"primaryKey" json:"id"`
	Nome               string   `gorm:"size:200;not null;index" json:"nome"`
	Codigo             string   `gorm:"size:50;uniqueIndex;not null" json:"codigo"`
	Descricao          *string  `gorm:"type:text" json:"descricao"`
	Tipo               string   `gorm:"size:30;default:'Preventiva'" json:"tipo"`
	Periodicidade      string   `gorm:"size:30;default:'Mensal'" json:"periodicidade"`
	DiasPersonalizado  *int     `gorm:"column:dias_personalizado" json:"dias_personalizado"`
	TempoEstimadoHoras *float64 `gorm:"column:tempo_estimado_horas;type:numeric(5,2)" json:"tempo_estimado_horas"`
	Criticidade        string   `gorm:"size:30;default:'Média'" json:"criticidade"`
	Prioridade         string   `gorm:"size:30;default:'Média'" json:"prioridade"`
	Ativo              bool     `gorm:"default:true" json:"ativo"`

	ResponsavelID  *uint `gorm:"column:responsavel_id" json:"responsavel_id"`
	DepartamentoID *uint `gorm:"column:departamento_id" json:"departamento_id"`
	CategoriaID    *uint `gorm:"column:categoria_id" json:"categoria_id"`

	DataCriacao        time.Time  `gorm:"column:data_criacao;default:CURRENT_TIMESTAMP" json:"data_criacao"`
	DataUltimaExecucao *time.Time `gorm:"column:data_ultima_execucao" json:"data_ultima_execucao"`
	ProximaExecucao    time.Time  `gorm:"column:proxima_execucao;not null" json:"proxima_execucao"`

	Responsavel  *User                  `gorm:"foreignKey:ResponsavelID" json:"responsavel,omitempty"`
	Departamento *Departamento          `gorm:"foreignKey:DepartamentoID" json:"departamento,omitempty"`
	Categoria    *AssetCategory         `gorm:"foreignKey:CategoriaID" json:"categoria,omitempty"`
	Assets       []MaintenancePlanAsset `gorm:"foreignKey:PlanID" json:"assets,omitempty"`
	Checklists   []MaintenanceChecklist `gorm:"foreignKey:PlanID" json:"checklists,omitempty"`
}

func (MaintenancePlan) TableName() string { return "maintenance_plans" }

type MaintenancePlanAsset struct {
	ID      uint `gorm:"primaryKey" json:"id"`
	PlanID  uint `gorm:"column:plan_id;not null" json:"plan_id"`
	AssetID uint `gorm:"column:asset_id;not null" json:"asset_id"`

	Plan  *MaintenancePlan `gorm:"foreignKey:PlanID" json:"plan,omitempty"`
	Asset *Asset           `gorm:"foreignKey:AssetID" json:"asset,omitempty"`
}

func (MaintenancePlanAsset) TableName() string { return "maintenance_plan_assets" }

type MaintenanceChecklist struct {
	ID     uint   `gorm:"primaryKey" json:"id"`
	PlanID uint   `gorm:"column:plan_id;not null" json:"plan_id"`
	Nome   string `gorm:"size:200;not null" json:"nome"`
	Ordem  int    `gorm:"default:0" json:"ordem"`

	Plan  *MaintenancePlan           `gorm:"foreignKey:PlanID" json:"plan,omitempty"`
	Items []MaintenanceChecklistItem `gorm:"foreignKey:ChecklistID" json:"items,omitempty"`
}

func (MaintenanceChecklist) TableName() string { return "maintenance_checklists" }

type MaintenanceChecklistItem struct {
	ID          uint   `gorm:"primaryKey" json:"id"`
	ChecklistID uint   `gorm:"column:checklist_id;not null" json:"checklist_id"`
	Descricao   string `gorm:"type:text;not null" json:"descricao"`
	Obrigatorio bool   `gorm:"default:true" json:"obrigatorio"`
	Ordem       int    `gorm:"default:0" json:"ordem"`
	RequerFoto  bool   `gorm:"column:requer_foto;default:false" json:"requer_foto"`

	Checklist *MaintenanceChecklist `gorm:"foreignKey:ChecklistID" json:"checklist,omitempty"`
}

func (MaintenanceChecklistItem) TableName() string { return "maintenance_checklist_items" }

type MaintenanceOrder struct {
	ID                  uint    `gorm:"primaryKey" json:"id"`
	Numero              string  `gorm:"size:50;uniqueIndex;not null" json:"numero"`
	PlanID              *uint   `gorm:"column:plan_id" json:"plan_id"`
	AssetID             *uint   `gorm:"column:asset_id" json:"asset_id"`
	InfraPredialServico *string `gorm:"column:infra_predial_servico;size:255" json:"infra_predial_servico"`

	TecnicoID     *uint `gorm:"column:tecnico_id" json:"tecnico_id"`
	SolicitanteID *uint `gorm:"column:solicitante_id" json:"solicitante_id"`

	Status      string `gorm:"size:30;default:'Aberta'" json:"status"`
	Prioridade  string `gorm:"size:30;default:'Média'" json:"prioridade"`
	Criticidade string `gorm:"size:30;default:'Média'" json:"criticidade"`
	Tipo        string `gorm:"size:30;default:'Preventiva'" json:"tipo"`

	DataAbertura      time.Time  `gorm:"column:data_abertura;default:CURRENT_TIMESTAMP" json:"data_abertura"`
	DataAgendada      *time.Time `gorm:"column:data_agendada" json:"data_agendada"`
	DataInicio        *time.Time `gorm:"column:data_inicio" json:"data_inicio"`
	DataPausa         *time.Time `gorm:"column:data_pausa" json:"data_pausa"`
	DataConclusao     *time.Time `gorm:"column:data_conclusao" json:"data_conclusao"`
	TempoTotalMinutos *int       `gorm:"column:tempo_total_minutos" json:"tempo_total_minutos"`

	Observacoes *string `gorm:"type:text" json:"observacoes"`
	Solucao     *string `gorm:"type:text" json:"solucao"`

	CustoTotal *float64 `gorm:"column:custo_total;type:numeric(10,2);default:0" json:"custo_total"`

	ServiceTicketID *uint `gorm:"column:service_ticket_id" json:"service_ticket_id"`

	Plan       *MaintenancePlan       `gorm:"foreignKey:PlanID" json:"plan,omitempty"`
	Asset      *Asset                 `gorm:"foreignKey:AssetID" json:"asset,omitempty"`
	Tecnico    *User                  `gorm:"foreignKey:TecnicoID" json:"tecnico,omitempty"`
	Executions []MaintenanceExecution `gorm:"foreignKey:OrderID" json:"executions,omitempty"`
	Materials  []MaintenanceMaterial  `gorm:"foreignKey:OrderID" json:"materials,omitempty"`
	Photos     []MaintenancePhoto     `gorm:"foreignKey:OrderID" json:"photos,omitempty"`
	History    []MaintenanceHistory   `gorm:"foreignKey:OrderID" json:"history,omitempty"`
}

func (MaintenanceOrder) TableName() string { return "maintenance_orders" }

type MaintenanceExecution struct {
	ID              uint       `gorm:"primaryKey" json:"id"`
	OrderID         uint       `gorm:"column:order_id;not null" json:"order_id"`
	ChecklistItemID uint       `gorm:"column:checklist_item_id;not null" json:"checklist_item_id"`
	Concluido       bool       `gorm:"default:false" json:"concluido"`
	Observacao      *string    `gorm:"type:text" json:"observacao"`
	DataExecucao    *time.Time `gorm:"column:data_execucao" json:"data_execucao"`
	ExecutadoPorID  *uint      `gorm:"column:executado_por_id" json:"executado_por_id"`

	Order         *MaintenanceOrder         `gorm:"foreignKey:OrderID" json:"order,omitempty"`
	ChecklistItem *MaintenanceChecklistItem `gorm:"foreignKey:ChecklistItemID" json:"checklist_item,omitempty"`
	ExecutadoPor  *User                     `gorm:"foreignKey:ExecutadoPorID" json:"executado_por,omitempty"`
}

func (MaintenanceExecution) TableName() string { return "maintenance_executions" }

type MaintenanceMaterial struct {
	ID            uint    `gorm:"primaryKey" json:"id"`
	OrderID       uint    `gorm:"column:order_id;not null" json:"order_id"`
	ProductID     *uint   `gorm:"column:product_id" json:"product_id"`
	Produto       string  `gorm:"size:200;not null" json:"produto"`
	Quantidade    float64 `gorm:"type:numeric(10,2);not null" json:"quantidade"`
	ValorUnitario float64 `gorm:"column:valor_unitario;type:numeric(10,2);not null" json:"valor_unitario"`
	ValorTotal    float64 `gorm:"column:valor_total;type:numeric(10,2);not null" json:"valor_total"`
	Observacao    *string `gorm:"type:text" json:"observacao"`

	Order *MaintenanceOrder `gorm:"foreignKey:OrderID" json:"order,omitempty"`
}

func (MaintenanceMaterial) TableName() string { return "maintenance_materials" }

type MaintenancePhoto struct {
	ID             uint      `gorm:"primaryKey" json:"id"`
	OrderID        uint      `gorm:"column:order_id;not null" json:"order_id"`
	ExecutionID    *uint     `gorm:"column:execution_id" json:"execution_id"`
	Tipo           string    `gorm:"size:30;default:'Durante'" json:"tipo"`
	CaminhoArquivo string    `gorm:"column:caminho_arquivo;size:255;not null" json:"caminho_arquivo"`
	Descricao      *string   `gorm:"type:text" json:"descricao"`
	DataUpload     time.Time `gorm:"column:data_upload;default:CURRENT_TIMESTAMP" json:"data_upload"`
	UploadPorID    *uint     `gorm:"column:upload_por_id" json:"upload_por_id"`

	Order     *MaintenanceOrder     `gorm:"foreignKey:OrderID" json:"order,omitempty"`
	Execution *MaintenanceExecution `gorm:"foreignKey:ExecutionID" json:"execution,omitempty"`
	UploadPor *User                 `gorm:"foreignKey:UploadPorID" json:"upload_por,omitempty"`
}

func (MaintenancePhoto) TableName() string { return "maintenance_photos" }

type MaintenanceHistory struct {
	ID             uint      `gorm:"primaryKey" json:"id"`
	OrderID        uint      `gorm:"column:order_id;not null" json:"order_id"`
	Acao           string    `gorm:"size:100;not null" json:"acao"`
	Descricao      string    `gorm:"type:text;not null" json:"descricao"`
	UsuarioID      *uint     `gorm:"column:usuario_id" json:"usuario_id"`
	DataHora       time.Time `gorm:"column:data_hora;default:CURRENT_TIMESTAMP" json:"data_hora"`
	StatusAnterior *string   `gorm:"column:status_anterior;size:50" json:"status_anterior"`
	StatusNovo     *string   `gorm:"column:status_novo;size:50" json:"status_novo"`

	Order   *MaintenanceOrder `gorm:"foreignKey:OrderID" json:"order,omitempty"`
	Usuario *User             `gorm:"foreignKey:UsuarioID" json:"usuario,omitempty"`
}

func (MaintenanceHistory) TableName() string { return "maintenance_history" }

type MaintenanceNotification struct {
	ID          uint      `gorm:"primaryKey" json:"id"`
	OrderID     *uint     `gorm:"column:order_id" json:"order_id"`
	PlanID      *uint     `gorm:"column:plan_id" json:"plan_id"`
	UsuarioID   uint      `gorm:"column:usuario_id;not null" json:"usuario_id"`
	Tipo        string    `gorm:"size:50;not null" json:"tipo"`
	Mensagem    string    `gorm:"type:text;not null" json:"mensagem"`
	Lida        bool      `gorm:"default:false" json:"lida"`
	DataCriacao time.Time `gorm:"column:data_criacao;default:CURRENT_TIMESTAMP" json:"data_criacao"`

	Order   *MaintenanceOrder `gorm:"foreignKey:OrderID" json:"order,omitempty"`
	Plan    *MaintenancePlan  `gorm:"foreignKey:PlanID" json:"plan,omitempty"`
	Usuario *User             `gorm:"foreignKey:UsuarioID" json:"usuario,omitempty"`
}

func (MaintenanceNotification) TableName() string { return "maintenance_notifications" }

type CustomMaintenanceType struct {
	ID        uint      `gorm:"primaryKey" json:"id"`
	Nome      string    `gorm:"size:200;not null;uniqueIndex" json:"nome"`
	Descricao *string   `gorm:"type:text" json:"descricao"`
	CriadoEm  time.Time `gorm:"column:criado_em;default:CURRENT_TIMESTAMP" json:"criado_em"`
}

func (CustomMaintenanceType) TableName() string { return "custom_maintenance_types" }
